File: classical_ann_cpp/src/file_parser.cpp
#include "file_parser.hpp"
#include <fstream>
#include <iostream>
#include <sstream>
#include <cmath>

using namespace std;

static int32_t reverse_int(int32_t i) {
    unsigned char c1 = i & 255;
    unsigned char c2 = (i >> 8) & 255;
    unsigned char c3 = (i >> 16) & 255;
    unsigned char c4 = (i >> 24) & 255;
    return ((int32_t)c1 << 24) +
           ((int32_t)c2 << 16) +
           ((int32_t)c3 << 8) +
           c4;
}

Dataset<uint8_t> read_mnist(const string& path) {
    Dataset<uint8_t> ds;
    ifstream file(path, ios::binary);

    if (!file.is_open()) {
        cerr << "Cannot open MNIST file: " << path << "\n";
        return ds;
    }

    int32_t magic_number, num_images, num_rows, num_cols;

    file.read(reinterpret_cast<char*>(&magic_number), sizeof(magic_number));
    file.read(reinterpret_cast<char*>(&num_images), sizeof(num_images));
    file.read(reinterpret_cast<char*>(&num_rows), sizeof(num_rows));
    file.read(reinterpret_cast<char*>(&num_cols), sizeof(num_cols));

    magic_number = reverse_int(magic_number);
    num_images   = reverse_int(num_images);
    num_rows     = reverse_int(num_rows);
    num_cols     = reverse_int(num_cols);

    ds.dim = num_rows * num_cols;
    ds.data.reserve(num_images);

    for (int i = 0; i < num_images; ++i) {
        vector<uint8_t> image(ds.dim);
        file.read(reinterpret_cast<char*>(image.data()), ds.dim);
        ds.data.push_back(move(image));
    }

    return ds;
}

Dataset<float> read_sift(const string& path) {
    Dataset<float> ds;
    ifstream file(path, ios::binary);

    if (!file.is_open()) {
        cerr << "Cannot open SIFT file: " << path << "\n";
        return ds;
    }

    while (true) {
        int dim;
        if (!file.read(reinterpret_cast<char*>(&dim), sizeof(int))) break;

        vector<float> vec(dim);
        file.read(reinterpret_cast<char*>(vec.data()), dim * sizeof(float));
        if (!file) break;

        float norm = 0.0f;
        for (float v : vec) norm += v * v;
        norm = sqrt(norm);
        if (norm > 0)
            for (float& v : vec) v /= norm;

        ds.data.push_back(move(vec));
        ds.dim = dim;
    }

    return ds;
}

Dataset<float> read_bio(const string& path, int dim) {
    Dataset<float> ds;
    ds.dim = dim;

    ifstream file(path, ios::binary);
    if (!file.is_open()) {
        cerr << "Cannot open BIO embedding file: " << path << "\n";
        return ds;
    }

    while (true) {
        vector<float> vec(dim);
        file.read(reinterpret_cast<char*>(vec.data()),
                  dim * sizeof(float));
        if (!file) break;
        ds.data.push_back(move(vec));
    }

    return ds;
}

vector<SearchResult> load_ground_truth(
    const string& filename,
    double& overall_tTrue,
    int max_queries
) {
    ifstream fin(filename);
    vector<SearchResult> gt;
    overall_tTrue = 0.0;

    if (!fin.is_open()) {
        cerr << "Error opening ground truth file: " << filename << endl;
        return gt;
    }

    string line;
    SearchResult current;
    int query_count = 0;

    while (getline(fin, line)) {
        if (line.rfind("Query:", 0) == 0) {
            if (!current.indices.empty()) {
                gt.push_back(current);
                query_count++;
                if (max_queries > 0 && query_count >= max_queries) break;
                current = SearchResult();
            }
        }
        else if (line.rfind("Nearest neighbor-", 0) == 0) {
            int idx;
            istringstream(line.substr(line.find(":") + 1)) >> idx;
            current.indices.push_back(idx);
        }
        else if (line.rfind("distanceTrue:", 0) == 0) {
            float dist;
            istringstream(line.substr(line.find(":") + 1)) >> dist;
            current.dists.push_back(dist);
        }
        else if (line.rfind("tTrueAverage:", 0) == 0) {
            istringstream(line.substr(line.find(":") + 1)) >> current.time;
        }
        else if (line.rfind("Overall tTrueAverage:", 0) == 0) {
            istringstream(line.substr(line.find(":") + 1)) >> overall_tTrue;
        }
    }

    if (!current.indices.empty())
        gt.push_back(current);

    return gt;
}

File: neural_partition_ann/src/file_parser.py
# file_parser.py

import struct
import numpy as np
from pathlib import Path
import torch
from src.models import MLPClassifier
from typing import List, Union, Tuple, Dict

from src.custom_types import Dataset, SearchResult


def read_mnist(path: Union[str, Path]) -> Dataset:
    """Read MNIST binary file and return Dataset with uint8 vectors."""
    ds = Dataset(data=[])
    path = Path(path)

    with path.open('rb') as f:
        magic_number = struct.unpack('>I', f.read(4))[0]
        if magic_number != 2051:
            raise ValueError(f"Invalid MNIST magic number {magic_number}")

        num_images = struct.unpack('>I', f.read(4))[0]
        num_rows = struct.unpack('>I', f.read(4))[0]
        num_cols = struct.unpack('>I', f.read(4))[0]

        dim = num_rows * num_cols
        ds.dim = dim

        for _ in range(num_images):
            image = np.frombuffer(f.read(dim), dtype=np.uint8)
            ds.data.append(image)

    return ds


def read_sift(path: Union[str, Path]) -> Dataset:
    """Read SIFT binary file and return Dataset with float vectors (L2 normalized)."""
    ds = Dataset(data=[])
    path = Path(path)

    with path.open('rb') as f:
        while True:
            dim_bytes = f.read(4)
            if len(dim_bytes) < 4:
                break
            dim = struct.unpack('i', dim_bytes)[0]
            vec_bytes = f.read(dim * 4)
            if len(vec_bytes) < dim * 4:
                break
            vec = np.frombuffer(vec_bytes, dtype=np.float32).copy()
            norm = np.linalg.norm(vec)
            if norm > 0:
                vec /= norm
            ds.data.append(vec)
            ds.dim = dim

    return ds


def read_bio(path: Union[str, Path]) -> Dataset:
    """Read BIO binary file and return Dataset with float vectors."""
    ds = Dataset(data=[])
    path = Path(path)
    dim: int = 320 # Fixed dimension for BIO dataset
    data = np.fromfile(path, dtype=np.float32)

    vectors = data.reshape(-1, dim)

    for v in vectors:
        ds.data.append(v)

    ds.dim = dim
    return ds


def load_ground_truth(filename: Union[str, Path], max_queries: int = -1) -> Tuple[List[SearchResult], float]:
    """Load ground truth file into list of SearchResult objects and overall tTrueAverage."""
    filename = Path(filename)
    gt: List[SearchResult] = []
    overall_tTrue = 0.0
    current = SearchResult(indices=[], dists=[], time=0.0)
    query_count = 0

    with filename.open('r') as f:
        for line in f:
            line = line.strip()
            if line.startswith("Query:"):
                if current.indices:
                    gt.append(current)
                    query_count += 1
                    if 0 < max_queries <= query_count:
                        break
                    current = SearchResult(indices=[], dists=[], time=0.0)
            elif line.startswith("Nearest neighbor-"):
                idx = int(line.split(":")[1].strip())
                current.indices.append(idx)
            elif line.startswith("distanceTrue:"):
                dist = float(line.split(":")[1].strip())
                current.dists.append(dist)
            elif line.startswith("tTrueAverage:"):
                current.time = float(line.split(":")[1].strip())
            elif line.startswith("Overall tTrueAverage:"):
                overall_tTrue = float(line.split(":")[1].strip())

    if current.indices:
        gt.append(current)

    return gt, overall_tTrue


def load_model(model_path: str, device: torch.device) -> Tuple[MLPClassifier, int, int]:
    """Load trained MLP model from checkpoint (reconstruct exact architecture)."""

    try:
        checkpoint = torch.load(model_path, map_location=device, weights_only=False)
    except TypeError:
        checkpoint = torch.load(model_path, map_location=device)

    d = checkpoint['d']
    m = checkpoint['m']

    # Saved architecture params, fallback to defaults
    nodes = checkpoint.get('nodes', 64)
    layers = checkpoint.get('layers', 3)

    # Construct the model
    model = MLPClassifier(d_in=d, n_out=m, hidden=nodes, layers=layers).to(device)

    # Load weights
    model.load_state_dict(checkpoint['model_state'])
    model.eval()

    return model, d, m


def load_inverted_file(invfile_path: str) -> Dict[int, List[int]]:
    """Load inverted file (mapping from partition -> list of data indices)."""
    data = np.load(invfile_path, allow_pickle=True)
    inverted = {}
    for key in data.files:
        try:
            inverted[int(key)] = list(data[key])
        except Exception as e:
            print(f"Warning: could not parse key {key}: {e}")
    return inverted